### PR TITLE
util: Replace boost sleep with std sleep

### PR DESCRIPTION
**Pull request by maflcko**
We don't use the interruptible feature of boost's sleep anywhere, so replace it with the sleep in `std::thread`

### CODE REVIEW

File: src/util/time.cpp [hunk 1]
```diff
@@ -53,41 +59,22 @@ int64_t GetMockTime()
...
 
 void MilliSleep(int64_t n)
 {
-
-/**
- * Boost's sleep_for was uninterruptible when backed by nanosleep from 1.50
- * until fixed in 1.52. Use the deprecated sleep method for the broken case.
- * See: https://svn.boost.org/trac/boost/ticket/7238
- */
-#if defined(HAVE_WORKING_BOOST_SLEEP_FOR)
```
Comment (laanwj) [inline src/util/time.cpp]: oh wow, can we finally get rid of this abomination?
feel free to delete the detection of this as well from the build system!

**Comment by MarcoFalke** [comment]
|                  | before | after  |
|------------------|--------|--------|
| compile time (s) | 8.63   | 2.61   |
| compile rss (kB) | 446368 | 101168 |
| run time (ns)    | 85     | 25     |

Source:

```
make
./src/bench/bench_bitcoin --filter=BenchTime.*
```

File: src/util/time.cpp (hunk 1 above)
Comment (MarcoFalke) [reply src/util/time.cpp]: Good idea! Done.

**Comment by MarcoFalke** [comment]
Removed unused configure checks

**Comment by sipa** [comment]
Are the std:: versions compatible with our use of boost threadgroup interrupts?

**Comment by laanwj** [comment]
> Are the std:: versions compatible with our use of boost threadgroup interrupts?

Oh crap I knew there was a reason we hadn't got rid of the boost::sleep madness yet :cry:

**Comment by practicalswift** [comment]
> Hopefully that has no observable effects other than improved code clarity, faster compilation and faster run time.

There is an very observable effect of this PR on 32-bit systems:

Prior to this PR `bitcoind` would shut down with an assertion failure when the `time_t` wraparound occurs on 32-bit systems.

After this PR `bitcoind` will keep running but with negative time returned from these functions :-)

**Comment by practicalswift** [comment]
Perhaps we should add a test making sure `bitcoind` on 32-bit systems behave the way we want it to -- which I assume is immediate shutdown -- at time of the 32-bit `time_t` wraparound event.

**Comment by laanwj** [comment]
How do other programs handle this? I though Linux switched to 64 bit timer values on 32-bit systems internally a while ago, but I see on ARM `time_t` is still 32 bit.

**Comment by MarcoFalke** [comment]
> After this PR bitcoind will keep running but with negative time returned from these functions :-)

Can you please explain a bit more why this happens. Be reminded that the predefined `std::chrono::duration` use an underlying integer type that is large enough (for the y2038 to be of no problem).

**Comment by practicalswift** [comment]
> > After this PR bitcoind will keep running but with negative time returned from these functions :-)
> 
> Can you please explain a bit more why this happens. Be reminded that the predefined `std::chrono::duration`s use an underlying integer type that is large enough (for year 2038 to be of no problem).

As long as the syscall used returns 32-bit time on 32-bit systems this cannot be solved no matter what integer types are used in the standard library, no?

**Comment by MarcoFalke** [comment]
Yeah, and the you can't run the underlying system anyway. There is no need to try to figure out how to make Bitcoin Core work on a broken system.

**Comment by practicalswift** [comment]
@MarcoFalke Perhaps we should start thinking about a deprecation plan for 32-bit systems (see #16096)? Even if this is 19 years away software tends to live longer than originally anticipated :-)

**Comment by MarcoFalke** [comment]
I've removed the changes on the time getters, so that this doesn't drift off into a discussion about the y2038 problem

**Comment by practicalswift** [comment]
ACK fae86c38bca5c960462e53975314a0749db5d17d -- patch looks correct

**Comment by sipa** [comment]
Concept and code review ACK fae86c38bca5c960462e53975314a0749db5d17d

**Review by fanquake — APPROVED** [review]
ACK fae86c38bca5c960462e53975314a0749db5d17d - note that an instance of `DHAVE_WORKING_BOOST_SLEEP_FOR` was missed in the [linter](https://github.com/bitcoin/bitcoin/blob/master/test/lint/extended-lint-cppcheck.sh#L69), but that can be cleaned up later.

**Comment by ajtowns** [comment]
ACK fae86c38bca5c960462e53975314a0749db5d17d quick code review

**Comment by laanwj** [comment]
Postmortem ACK.

**Pull request merged**